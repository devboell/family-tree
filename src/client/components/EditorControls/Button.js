import styled from 'styled-components'


const Button = styled.button`
  padding: 3px 10px;
  margin-right: 20px;
  font-size: 12px;
  cursor: pointer;
`

export default Button
