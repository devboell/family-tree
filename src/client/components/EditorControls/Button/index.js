import styled from 'styled-components'


const Button = styled.button`
  padding: 3px 10px;
  margin-right: 20px;
  font-size: 12px;
`

export default Button
