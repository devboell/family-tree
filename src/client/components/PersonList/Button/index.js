import styled from 'styled-components'


const Wrapper = styled.button`
  width: 100%;
  background-color: rgb(76, 175, 80);
  border: none;
  color: white;
  padding: 15px 32px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  font-size: 12px;

  &:hover {
    background-color: gray;
  }
`

export default Wrapper
