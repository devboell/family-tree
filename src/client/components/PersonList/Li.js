import styled from 'styled-components'


const Li = styled.li`
  margin-bottom: 5px;
`

export default Li
