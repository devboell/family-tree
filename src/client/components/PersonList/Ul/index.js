import styled from 'styled-components'


const Ul = styled.ul`
  flex-basis: 20%;
  list-style: none;
  padding: 0px;
  margin: 0px 0px;
`

export default Ul
