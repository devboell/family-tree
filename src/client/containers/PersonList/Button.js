import styled, { css } from 'styled-components'

const hover = css`
  background-color: gray;
`

const hoverAndSelected = css`
  background-color: rgb(18, 106, 22);
  cursor: default;
`

const Button = styled.button`
  width: 100%;
  background-color: ${props => (
    props.isSelected
    ? 'rgb(18, 106, 22)'
    : 'rgb(76, 175, 80)'
  )};
  border: none;
  color: white;
  padding: 15px 32px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  font-size: 12px;
  outline: 0;
  cursor: pointer;

  &:hover {
    ${props => (
      props.isSelected
      ? hoverAndSelected
      : hover
    )};
  }
`
Button.displayName = 'Button'

export default Button
