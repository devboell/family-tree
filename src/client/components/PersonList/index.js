import React, { Component } from 'react'
import { PropTypes } from 'prop-types'
import Ul from './Ul'
import Li from './Li'
import Button from './Button'


class PersonList extends Component {
  componentDidMount() {
    const { persons, selectedPersonId, shouldPreselect, onSelectPersonId } = this.props

    if (persons.length &&
       selectedPersonId === 'no_selection' &&
       shouldPreselect) {
      onSelectPersonId(persons[0].id)
    }
  }

  render() {
    const {
      persons,
      selectedPersonId,
      onSelectPersonId,
    } = this.props

    return (
      <Ul>
        {persons.map(person =>
          <Li key={`person-${person.id}`}>
            <Button
              isSelected={person.id === selectedPersonId}
              onClick={() => onSelectPersonId(person.id)}
            >
              {person.name}
            </Button>
          </Li>,
        )}
      </Ul>
    )
  }
}

PersonList.propTypes = {
  persons: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  selectedPersonId: PropTypes.string.isRequired,
  shouldPreselect: PropTypes.bool.isRequired, // eslint-disable-line react/no-unused-prop-types
  onSelectPersonId: PropTypes.func.isRequired,
}

PersonList.defaultProps = {
  shouldPreselect: true,
}

export default PersonList
