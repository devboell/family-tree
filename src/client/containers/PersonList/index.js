import React, { Component } from 'react'
import { PropTypes } from 'prop-types'
import { connect } from 'react-redux'
import { compose, isEmpty, find } from 'lodash/fp'

import withPersons from './enhancers'
import { selectPerson, selectFirst, unselect } from './thunks'
import selectionModes from './constants'

import Wrapper from './Wrapper'
import ChildWrapper from './ChildWrapper'
import ListWrapper from './ListWrapper'
import EmptyListMessage from './EmptyListMessage'
import Ul from './Ul'
import Li from './Li'
import Button from './Button'

export class PersonList extends Component {
  componentWillReceiveProps(nextProps) {
    const {
      loading,
      persons,
      onSelectPerson,
      selectionMode,
    } = nextProps

    const shouldSelectFirst =
      !loading &&
      persons.length &&
      selectionMode === selectionModes.SELECT_FIRST

    if (shouldSelectFirst) {
      onSelectPerson(persons[0].id)
    }
  }

  render() {
    const {
      loading,
      persons,
      refetch,
      selectedPersonId,
      selectionMode,
      onSelectPerson,
      onSelectFirst,
      onUnselect,
      children,
    } = this.props

    if (loading) return <Wrapper>loading ...</Wrapper>

    return (
      <Wrapper>
        <ListWrapper>
          {isEmpty(persons)
            ? <EmptyListMessage>no persons found</EmptyListMessage>
            : <Ul>
              {persons.map(person =>
                <Li key={`person-${person.id}`}>
                  <Button
                    isSelected={person.id === selectedPersonId}
                    onClick={() => onSelectPerson(person.id)}
                  >
                    {person.name}
                  </Button>
                </Li>,
              )}
            </Ul>
          }
        </ListWrapper>

        <ChildWrapper>
          {React.cloneElement(
            React.Children.only(children),
            {
              personId: selectedPersonId,
              person: find({ id: selectedPersonId })(persons) || {},
              selectionMode,
              onSelectPerson,
              onSelectFirst,
              onUnselect,
              refetch,
              persons,
            },
          )}
        </ChildWrapper>
      </Wrapper>
    )
  }
}


PersonList.propTypes = {
  loading: PropTypes.bool.isRequired,
  persons: PropTypes.arrayOf(PropTypes.shape({})),
  refetch: PropTypes.func.isRequired,
  selectedPersonId: PropTypes.string.isRequired,
  selectionMode: PropTypes.string.isRequired,
  onSelectPerson: PropTypes.func.isRequired,
  onSelectFirst: PropTypes.func.isRequired,
  onUnselect: PropTypes.func.isRequired,
  children: PropTypes.node.isRequired,
}

PersonList.defaultProps = {
  persons: [],
}

export const mapStateToProps = state => ({
  selectedPersonId: state.personList.selectedPersonId,
  selectionMode: state.personList.selectionMode,
})


export const mapDispatchToProps = dispatch => ({
  onSelectPerson: id => dispatch(selectPerson(id)),
  onSelectFirst: () => dispatch(selectFirst()),
  onUnselect: () => dispatch(unselect()),
})

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withPersons,
)(PersonList)
