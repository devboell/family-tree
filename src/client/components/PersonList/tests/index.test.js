import React from 'react'
import PersonList from '../index'
import Button from '../Button'

const props = {
  persons: [
    {
      name: 'John',
    },
    {
      name: 'Mary',
    },
  ],
  onSelectPerson: jest.fn(),
}

it('PersonList component', () => {
  expect(shallow(<PersonList {...props} />)).toMatchSnapshot()
})

it('PersonList button click', () => {
  const wrapper = shallow(<PersonList {...props} />)
  wrapper.find(Button).at(0).simulate('click')
  expect(props.onSelectPerson).toHaveBeenCalled()
})
