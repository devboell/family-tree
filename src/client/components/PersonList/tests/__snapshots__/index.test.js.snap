// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonList component 1`] = `
<styled.ul>
  <styled.li>
    <styled.button
      onClick={[Function]}
    >
      John
    </styled.button>
  </styled.li>
  <styled.li>
    <styled.button
      onClick={[Function]}
    >
      Mary
    </styled.button>
  </styled.li>
</styled.ul>
`;
